src/dmd/env.d

```diff
@@ -0,0 +1,113 @@
+module dmd.env;
+
+import core.stdc.string;
+import core.sys.posix.stdlib;
+import dmd.globals;
+import dmd.root.array;
+import dmd.root.rmem;
+import dmd.utils;
+
+version (Windows)
+    private extern (C) int putenv(const char*) nothrow;
+
+/**
+Construct a variable from `name` and `value` and put it in the environment while saving
+the current value of the environment variable.
+Returns:
+    0 on success, non-zero on failure
+*/
+int putenvRestorable(const(char)[] name, const(char)[] value) nothrow
+{
+    auto var = LocalEnvVar.xmalloc(name, value);
+    const result = var.putenvRestorable();
+    var.xfree(result);
+    return result;
+}
+
+/// Holds a `VAR=value` string that can be put into the global environment.
+struct LocalEnvVar
+{
+    string nameValueCStr;        // The argument passed to `putenv`.
```

Comment:
Note it's called "nameValue" here, not "VARvalue"

Reply:
Note in the comment that it has a terminating 0 past the end of the string, not as part of the name.